lib/pymedphys/_interp/interp.py

```diff
@@ -0,0 +1,331 @@
+            or points_interp[:, i].max() > axis_known.max()
+        ):
+            raise ValueError(
+                f"""points_interp[:, {i}] must be within the range of axes_known[{i}]\n
+                ({points_interp[:, i].min()}, {points_interp[:, i].max()}) vs. ({axis_known.min()}, {axis_known.max()})"""
+            )
+
+    axes_known = tuple(np.array(axis, dtype=np.float64) for axis in axes_known)
+    values = np.array(values, dtype=np.float64)
+
+    return axes_known, values
+
+
+@nb.njit(parallel=True, fastmath=True, cache=True)
+def interp1d(axis_known, values, points_interp, extrap_fill_value=np.nan):
+    values_interp = np.zeros(points_interp.shape[0], dtype=np.float64)
+    diff = axis_known[1] - axis_known[0]
+
+    # pylint: disable=not-an-iterable
+    for i in nb.prange(points_interp.shape[0]):
+        xpi = points_interp[i, 0]
+
+        if not axis_known[0] <= xpi <= axis_known[-1]:
+            values_interp[i] = extrap_fill_value
+            continue
+
+        x1_idx = np.searchsorted(axis_known, xpi)
+        x0_idx = x1_idx - 1
+
+        if x0_idx < 0:
+            x0_idx = 0
+        if x1_idx >= axis_known.size:
+            x1_idx = axis_known.size - 1
+
+        wx = (xpi - axis_known[x0_idx]) / diff
+
+        values_interp[i] = values[x0_idx] * (1 - wx) + values[x1_idx] * wx
+
+    return values_interp
+
+
+@nb.njit(parallel=True, fastmath=True, cache=True)
+def interp2d(axes_known, values, points_interp, extrap_fill_value=np.nan):
+    values_interp = np.zeros((points_interp.shape[0]), dtype=np.float64)
+    diffs = np.zeros(2)
+    for i, axis in enumerate(axes_known):
+        diffs[i] = axis[1] - axis[0]
+    x, y = axes_known
+
+    # pylint: disable=not-an-iterable
+    for i in nb.prange(points_interp.shape[0]):
+        xpi, ypi = points_interp[i, 0], points_interp[i, 1]
+
+        if not x[0] <= xpi <= x[-1] or not y[0] <= ypi <= y[-1]:
+            values_interp[i] = extrap_fill_value
+            continue
+
+        # Find the indices of the surrounding grid points
+        x1_idx = np.searchsorted(x, xpi)
+        x0_idx = x1_idx - 1
+        y1_idx = np.searchsorted(y, ypi)
+        y0_idx = y1_idx - 1
+
+        if x0_idx < 0:
+            x0_idx = 0
+        if y0_idx < 0:
+            y0_idx = 0
+        if x1_idx >= x.size:
+            x1_idx = x.size - 1
+        if y1_idx >= y.size:
+            y1_idx = y.size - 1
+
+        c00 = values[x0_idx, y0_idx]
+        c01 = values[x0_idx, y1_idx]
+        c10 = values[x1_idx, y0_idx]
+        c11 = values[x1_idx, y1_idx]
+
+        wx = (xpi - x[x0_idx]) / diffs[0]
+        wy = (ypi - y[y0_idx]) / diffs[1]
+
+        c0 = c00 * (1 - wx) + c10 * wx
+        c1 = c01 * (1 - wx) + c11 * wx
+
+        values_interp[i] = c0 * (1 - wy) + c1 * wy
+
+    return values_interp
+
+
+@nb.njit(parallel=True, fastmath=True, cache=True)
+# pylint: disable=invalid-name
+def interp3d(axes_known, values, points_interp, extrap_fill_value=np.nan):
```

Comment:
Key would be to make a wrapper function that only needs the nb import for njit when the function is called for the first time. Approach:

```python
from functools import cache

@cache
def _get_interp3d():
    @nb.njit(parallel=True, fastmath=True, cache=True)
    # pylint: disable=invalid-name
    def _interp3d(axes_known, values, points_interp, extrap_fill_value=np.nan):
        ...

    return _interp3d

def interp3d(axes_known, values, points_interp, extrap_fill_value=np.nan):
    _interp3d = _get_interp3d()
    return _interp3d(axes_known=axes_known, values=values, points_interp=points_interp, extrap_fill_value=extrap_fill_value)
```